F5 router plugin

Miciah opened the pull request.

Miciah commented:
@pweil-, does this look OK?

pweil- commented:
@Miciah - this needs a pretty rebase to pick up the various PRs for routing that have landed.  Please take a look at the admin/router/router.go as well, this is the recommended way to creating routers for administrators.  Also, please be aware that https://github.com/openshift/origin/pull/2565 is about to land that will create a config object similar to what you are doing in this PR.

When you rebase can you squash this?  Thanks

Miciah commented:
@pweil-, I rebased Friday, but I neglected to check for new conflicts this morning.  I'll wait for #2565 to merge and then I'll rebase again.  Thanks!

Miciah commented:
@pweil-, I squashed, rebased, rebuilt, and verified that `oadm router --credentials=/root/os-certs/master/admin.kubeconfig --type=f5-router --latest-images --external-host=[redacted] --external-host-username=[redacted] --external-host-password=[redacted] --external-host-insecure=true` worked.  I had to edit rc/router-1 afterwards to change replicas from 0 to 1, but that appears to be a problem unrelated to this PR (see BZ1230005).

Let me know if I oversquashed—I wasn't sure whether you wanted a more logical breakdown of commits or just one big commit.

Does the approach in `pkg/cmd/admin/router/router.go` look OK? I followed the existing approach there, keeping the options generic and all in `RouterConfig`, rather than following the approach you suggested to me a while back for `pkg/cmd/infra/router/router.go` of creating a generic `routerConfig`, subordinating `templateRouterConfig` underneath, and adding an `f5RouterConfig` alongside.  Is the approach I took what you wanted?

I also have changed the F5 router plug-in and F5 client interfaces some to follow conventions from recent changes to the template router plug-in and HAProxy implementation.  The coupling and visibility conventions were not always obvious to me, so let me know if anything looks amiss.

pweil- commented:
thanks @Miciah will review (probably tonight).  

/cc @rajatchopra @ramr for extra eyes

Miciah commented:
Rebased (and rebuilt and briefly tested) again for a conflict with PR 3117.

pweil- commented on images/router/f5/pod.json:
```diff
@@ -0,0 +1,45 @@
+{
```
Is this used for anything?  I don't think we need it.

Miciah replied:
I can remove it; I snabbed it from https://github.com/pweil-/origin/commit/fa89f112c1263324de59b4661389a5c6606904fb

pweil- commented on pkg/cmd/admin/router/router.go:
```diff
@@ -255,18 +265,28 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg *
 			fmt.Fprintf(out, "password for stats user %s has been set to %s\n", cfg.StatsUsername, cfg.StatsPassword)
 		}
 
+		externalHostInsecure := "false"
```
you can use strconv.FormatBool and remove this code

Miciah replied:
I'm following the style from 15 lines above.  Should both use `strconv.FormatBool`, or am I missing some subtlety?

pweil- commented on images/router/f5/Dockerfile:
```diff
@@ -0,0 +1,11 @@
+ADD bin/openshift /usr/bin/openshift
+RUN ln -s /usr/bin/openshift /usr/bin/openshift-router
+
+ENTRYPOINT ["/usr/bin/openshift-router", "--type=f5"]
```
this requires a change in the haproxy dockerfile as well, right?

Miciah replied:
I don't understand the question.  This image is built from `openshift/origin-base` rather than from `openshift/origin-haproxy-router-base` so I don't understand why it would require a change to the haproxy Dockerfile.

pweil- replied:
I'm referring to passing of the `--type`, the haproxy entrypoint now needs to pass `--type=template` if I understand correctly

Miciah replied:
I see what you mean.  An alternative would be to make `template` the default value for `--type`, but it would be cleaner to modify the haproxy router's `Dockerfile`, so I'll do that.  Good catch.

pweil- commented on pkg/cmd/infra/router/router.go:
```diff
@@ -37,35 +54,75 @@ type templateRouterConfig struct {
+	routerTypeTemplate = "template"
+)
+
+func newEmptyConfig() *routerConfig {
```
godoc, por favor

kargakis replied:
Note that this is only necessary for exported vars, though it's good to have everything documented

pweil- commented on pkg/cmd/infra/router/router.go:
```diff
@@ -37,35 +54,75 @@ type templateRouterConfig struct {
 	StatsUsername      string
 }
 
+// f5RouterConfig is the config necessary to start an F5 router plugin
```
I get dinged all the time about my godoc not being complete sentences so I'm passing along the favor.  Do a sweep and make sure it is all formatted with //<space> and ends with a period if it is a method or type godoc.

Miciah replied:
That came from your skeleton at <https://github.com/pweil-/origin/commit/f057519d1ff91c4a34273c659e26148421ebe872>.  *grin*.  I'll fix it.

pweil- replied:
lol, that does not surprise me one bit.

pweil- commented on pkg/cmd/infra/router/router.go:
```diff
@@ -26,8 +27,24 @@ This command launches a router connected to your OpenShift master. The router li
+	Type string
+
+	//TemplateRouterConfig holds config items for template router plugins
+	TemplateRouterConfig *templateRouterConfig
```
not sure the purposes of making these a pointer if they will always be initialized in `newEmptyConfig`.  Incidentally, if they aren't pointers we don't need `newEmptyConfig`.  I don't have a preference, just curious about the decision

Miciah replied:
That came from you skeleton.  I'll depointerize them.

Miciah replied:
Unless I'm missing something, we still need to initialise the `Config` field with `clientcmd.NewConfig()`.  Are you thinking I should just use something like `cfg := &routerConfig{Config: clientcmd.NewConfig()}` in `NewCommandTemplateRouter` (which I need to rename) and let the other fields in `routerConfig` assume zero values?

pweil- replied:
You can leave the `newEmptyConfig` method, it doesn't hurt anything, but yes, they can assume zero values since they're always needed for the cmd flag parsing.

pweil- commented on pkg/cmd/infra/router/router.go:
```diff
@@ -37,35 +54,75 @@ type templateRouterConfig struct {
 		Long:  routerLong,
 		Run: func(c *cobra.Command, args []string) {
+			if !isValidType(cfg.Type) {
+				glog.Fatal("Invalid type specified.  Valid values are template, f5")
```
change to `Fatalf` and output the type that was passed and use your constant values for the valid types

Miciah replied:
(y)

pweil- commented on pkg/cmd/infra/router/router.go:
```diff
@@ -37,35 +54,75 @@ type templateRouterConfig struct {
 }
 
+// f5RouterConfig is the config necessary to start an F5 router plugin
+type f5RouterConfig struct {
```
Will the f5 router not support using a wildcard certificate for auto-generated route names?  I see that `DefaultCertificate` as left in the `templateRouterConfig` but it seems like it should be relevant to both implementations.

Miciah replied:
It's not there in this PR.

pweil- replied:
Ok, if the intention is to support it longer term let's move the field up to the generic config so we're prepped for it and create a follow up issue to track it.

Miciah replied:
Will do.

pweil- commented on pkg/cmd/infra/router/router.go:
```diff
@@ -37,35 +54,75 @@ type templateRouterConfig struct {
+			var err error
+
+			switch cfg.Type {
+			case "f5":
```
use your constants

Miciah replied:
(y)

pweil- commented on pkg/cmd/infra/router/router.go:
```diff
@@ -37,35 +54,75 @@ type templateRouterConfig struct {
+				plugin, err = makeF5Plugin(cfg)
+			case "template":
+				plugin, err = makeTemplatePlugin(cfg)
+			}
```
default case can cover the if statement below with the `//should never get here`

Miciah replied:
OK, I'll add a `default:` case that assigns a non-nil error message to `err` for the `if` after the `switch`.

pweil- commented on pkg/cmd/infra/router/router.go:
```diff
@@ -74,47 +131,92 @@ func NewCommandTemplateRouter(name string) *cobra.Command {
+	}
+
+	if cfg.F5RouterConfig.Insecure == true {
+		glog.Warning("Strict certificate verification is *DISABLED*")
```
Seems like this belongs in the log output of the plugin itself.

Miciah replied:
I guess the principle is along the lines that validation of the parameters passed to the plug-in belongs here but additional logging output belongs in the plug-in itself? I'll move the warning to `newF5LTM`.

pweil- commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,555 @@
+	url            string
+	httpStatusCode int
+	err            error
+	JsonStatusCode int     `json:"code"`
```
Why the json definition on these two fields.  Also, I didn't see this field being used anywhere

Miciah replied:
`JsonStatusCode` and `Message` are populated from the JSON blob in the HTTP response body whereas `httpStatusCode` is assigned the code in the HTTP response header.

Fair point that there are no current users of `JsonStatusCode`.  I could add validations to check that `JsonStatusCode` matches `httpStatusCode`, but there probably isn't much value in doing so, so I'll just remove `JsonStatusCode`.

pweil- commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,555 @@
+}
+
+type f5LTM struct {
+	host     string
```
you can add the `f5LTMCfg` type here and not have to define these fields

Miciah replied:
I did it this way to parallel `templateRouter` and `templateRouterCfg`, although there `templateRouterCfg` is not strictly a subset of `templateRouter`, so following the same approach is misguided.  I'll follow your suggestion here.

pweil- commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,555 @@
+
+	req, err := http.NewRequest(verb, url, payload)
+	if err != nil {
+		errorResult.err = fmt.Errorf("http.NewRequest failed: %v", err)
```
Should we include info about the request in these messages?  Logging the verb and url might help troubleshooting in production.

pweil- replied:
If I just haven't read far enough, ignore.

Miciah replied:
Yeah, that's what the big fancy `F5Error` structure and its `Error` method give you—that seemed like the golangish way to do it.  Is it failing to log something that it could be logging, or is the approach overengineered or otherwise bad?

pweil- replied:
No, that's fine.  I just missed the fact that errorResult is F5Error.  Looks good to me

Miciah commented on pkg/cmd/infra/router/router.go:
```diff
@@ -37,35 +54,75 @@ type templateRouterConfig struct {
+}
+
 // NewCommndTemplateRouter provides CLI handler for the template router backend
 func NewCommandTemplateRouter(name string) *cobra.Command {
```
I should rename this function.

pweil- commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,555 @@
+		return err
+	}
+	if hasMember {
+		return fmt.Errorf("Pool %s already has member %s.\n", poolName, member)
```
Is this really a failure case or can we succeed and carry on since there is no work to do?

Miciah replied:
The plug-in should check whether the member is in the pool before it tries to add it.  There is a logic error if we hit this case.  If you believe it would be better to log and continue though, I can make that change.

pweil- replied:
Agree on checking.  I haven't got to the HandleEndpoints method for an in depth look but I think you've covered my concern in that method.  It will just log the error and continue adding the other endpoints.

pweil- commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,555 @@
+		Name:            "0",
+		CaseInsensitive: true,
+		External:        true,
+		HttpHost:        true,
```
It may be my F5 ignorance, but how does this support tcp passthrough routes and the other route types like edge terminated and reencrypt?

Miciah replied:
Right now it supports only passthrough.

pweil- replied:
Do you mean only unsecure http routes?  When I say passthrough I mean a tcp connection wrapped in TLS with SNI that may not be http requests.

Miciah replied:
Sorry, that's right.  This PR does not give you TLS.

pweil- commented on plugins/router/f5/plugin.go:
```diff
@@ -0,0 +1,183 @@
+
+		// The invocation of PoolHasMember has the side effect that it will
+		// initialise PoolMembers[poolName], which is used below, if necessary.
+		poolExists, err := p.F5Client.PoolExists(poolName)
```
We can't rely on the route being created before the endpoints event reaches the plugin.

Miciah replied:
Would it be fine if we created the pool if necessary in response to the endpoint event and only add the vhost routing rule (creating the pool then if necessary) when the route is created? We'd be creating redundant pools (e.g., for `kubernetes` and `kubernetes-ro`, as well as anything else that defines a service but no route), but it seems like the most sensible approach.

pweil- replied:
Yes, that's essentially what we do in the template router (it saves endpoints and only writes backends if it has routes+endpoints).  The alternate, more complicated approach would to be maintaining some internal state and only creating the pools when you have all the information necessary.

Miciah replied:
Maintaining internal state is suboptimal, so I'll go with the simpler approach.

pweil- commented on plugins/router/f5/plugin.go:
```diff
@@ -0,0 +1,183 @@
+			return err
+		}
+
+		err = p.F5Client.AddPoolRoute(poolName, hostname)
```
Routes allow you to add a path component for routing as well.

Miciah replied:
OK, F5 supports that; it should be an easy change.  Would we need to match on both the vhost and the path?

pweil- replied:
Yes.  I would assume that the rules in F5 have some sort of specificity ordering ie. if I define route www.example.com then define www.example.com/test it will first look for a path based route and fall back to www.example.com if no matches are found.

pweil- commented:
@Miciah - first pass done.  I really like how the CRUD is done in the f5 portion, I can easily see us making that generic for all REST based routers.  

Please sweep for godoc in all the files.  Some notable things that I'm not sure are requirements or not for your initial example, you might want to confirm if you're on the hook for these:

1.  Support for passthrough
1.  Support for edge
1.  Support for reencrypt
1.  Path based routing

pweil- commented:
Oh, and unit tests.  Also, there is an integration test for routers that may help you test the use cases for your image.  It uses docker and you can specify the image it runs.

smarterclayton commented:
What's the status of this?

Miciah commented:
Working on the unit tests.

pweil- commented on plugins/router/f5/plugin.go:
```diff
@@ -0,0 +1,466 @@
+	}
+
+	// Name for the route in F5.
+	routename := fmt.Sprintf("openshift_route_%s_%s", route.Namespace, route.Name)
```
This pattern is used 8 times, make it a function.

Miciah replied:
Done.  It will be fixed in a follow-up commit.

pweil- commented on plugins/router/f5/plugin.go:
```diff
@@ -0,0 +1,466 @@
+		// here) and then adding the updated route (by falling through from the
+		// watch.Deleted case to the watch.Added case).
+		modifying = true
+		fallthrough
```
This is a bit hard to read and relies on case ordering.  I think it may be easier to follow by factoring the logic out to functions and doing something like

```go
switch type {
  case Mod: 
      delete()
      add()
  case Delete:
      delete()
  case Add:
      add()
}

Miciah replied:
This is refactored in the latest update.

pweil- commented on plugins/router/f5/plugin.go:
```diff
@@ -0,0 +1,466 @@
+		// Start with the routes because we cannot delete the pool until we delete
+		// any associated profiles and rules.
+
+		secureRouteExists, err := p.F5Client.SecureRouteExists(routename)
```
Do we need to make these calls every time for every route or should we base this (and the call below) on the TLS section of the route?  I imagine it doesn't save a whole lot unless there are lots of updates going on.

Miciah replied:
The primary motivation for doing it this way is that it enables us to re-use the code for `watch.Modified` where we don't know what type of route it was before the event, only what type of route it is now.

These checks generally should be using only information that the router is already storing, so it shouldn't result in significant overhead.  Avoiding these checks would mean that we couldn't use the fallthrough approach or the approach you suggest above of handling `watch.Modified` by re-using the same `delete()` and `add()` that we would use for `watch.Deleted` and `watch.Added`—unless we added a flag or a several flags to `delete()` to indicate which types of route to delete, which would be slightly hairy but certainly no more so than, say, `deleteCertParts()` in `f5.go`...

pweil- commented on pkg/cmd/infra/router/router.go:
```diff
@@ -26,28 +27,104 @@ This command launches a router connected to your OpenShift master. The router li
+	routerTypeTemplate = "template"
+)
+
+// NewCommandRouter provides a CLI handler for the template router backend.
```
minor nit: this is no longer just for the template router

Miciah replied:
This will be fixed in a follow-up commit.

pweil- commented:
@Miciah - a couple of comments on the TLS implementation pieces.  It would be good, now that all the use cases are filled in, to run through the router integration test for this container.

Miciah commented:
@pweil-, thanks! I addressed the comments, although one of them (concerning the handling of `watch.Modified`) needs further discussion.  What do I do to run through the router integration test?

pweil- commented:
@Miciah thanks, updates are looking good.  To use the integration test you can take a look at `test/integration/router_test.go`.  Since you have an external instance to connect to you'll do the following

1.  Have your external instance running and reachable from where you're running the test
1.  modify `defaultRouterImage` to be your image
1.  modify `createAndStartRouter` to pass any arguments you need to your container if they do not default to good values
1.  run the integration test with docker to run your test

@smarterclayton if you want any other eyes on this I think now is a good time.  @rajatchopra @ramr?

Miciah commented:
@pweil-, sorry, it took me a little while to work through issues with the integration tests.  I shall proceed describe what I had to do, and convince you that I'm not doing anything naughty, with my latest commit.

• In the spirit of your suggestion, I modify the initialisation code in `createAndStartRouterContainer` to copy relevant environment variables into the router container environment to facilitate specifying F5 BIG-IP credentials and other configuration.  This seemed like the cleanest approach to facilitate automated test-runs with no code changes and minimal configuration effort to switch between template and F5 routers.

• The tests create routes and then connect to these routes.  The tests currently assume that the routes will be available at 0.0.0.0 (i.e., the local host), which is fine for the template router running on the same host where the tests run.  However, in the case of F5, these routes are on a distinct host from the host on which the tests run, and so it is necessary to permit this default to be overridden, which the latest commit on this PR allows via the `OPENSHIFT_ROUTE_ADDRESS` environment variable (as used by the new `getRouteAddress` function).

• The tests create endpoints that point back to a test webserver on the host on which the tests run.  The tests currently assume that this test webserver will be available at 0.0.0.0, which again (and for similar reasons) is fine for the template router, but not for F5.  The latest commit on this PR changes the tests to create the services using the actual address of the host (as determined by the new `GetDefaultLocalAddress` function), not simply 0.0.0.0.

• The tests currently check for a default route that connects with the OpenShift master.  I delete this check because I am not aware of any requirement for a default route, but if you prefer, we can keep the default route and discuss how it should be implemented in the context of F5 (perhaps a configuration option?).

• The websocket tests are a bit whack, and unfortunately so is the websocket client code.  First of all, the tests were using an IP address instead of a hostname in their aliases, which F5 BIG-IP rejects as an invalid servername (reasonably so, I believe), so I changed these routes to use a proper hostname (www.example.com).  I also fixed `getRoute` for the websocket case to set the the "Origin" header using a correct address (re-using the aforementioned `GetDefaultLocalAddress` function) instead of "localhost".  I additionally had to fixed `getRoute` to set the "Host" header the correct way for the websocket client code, whereupon I discovered that, unlike the HTTP client code, the websocket client code does not allow a "Host" header that is distinct from the address to which the client code attempts to connect—so if we are putting "www.example.com" in the "Host" header, the client will connect to "www.example.com".  As a workaround, I am manually adding an entry for www.example.com to `/etc/hosts`.  Should I document this workaround in a README or in a comment, or can you think of a better way to resolve the issue? Shall I file a report (if none is already field) regarding this shortcoming against the websocket client?

• Some tests do not clean up after themselves—they leave configuration cruft in F5 BIG-IP, which is rather inconsiderate and moreover causes consecutive test-runs to fail.  My latest commit fixes the tests to be more considerate, and also adds some sleeps at the ends of test cases so that the router will have time to update F5 BIG-IP before the deferred `cleanUp(dockerCli, routerId)` calls kills the router.

• The test cases do not give adequate time between setting up services and routes and attempting to connect to those routes.  I had to add or increase some sleeps so that F5 BIG-IP will be ready when the test cases try to connect to routes that it serves.

• While I was at it, I update the comment for `getRoute` because its behaviour is weird and non-intuitive in several respects.

With these changes (including the `/etc/hosts` workaround), `hack/test-integration-docker.sh` succeeds both still with the template router and now also with the F5 router (if I export appropriate settings for `OPENSHIFT_ROUTER_IMAGE`, `OPENSHIFT_ROUTE_ADDRESS`, `ROUTER_EXTERNAL_HOST_HOSTNAME`, `ROUTER_EXTERNAL_HOST_USERNAME`, `ROUTER_EXTERNAL_HOST_PASSWORD`, `ROUTER_EXTERNAL_HOST_INSECURE`, and `ROUTER_EXTERNAL_HOST_PRIVKEY`).

Let me know what you think!

ramr commented on pkg/cmd/admin/router/router.go:
```diff
@@ -135,7 +219,8 @@ func loadDefaultCert(file string) (string, error) {
 	return string(bytes), err
 }
```
Any reason to duplicate code within the 2 methods - they both look like the same code (its late so I might have missed something here). Would be better to just have one method called ```loadData``` or something to that effect which has the read data + error handling code and have the other 2 methods just be ```aliases``` effectively.

ramr commented on pkg/cmd/infra/router/router.go:
```diff
@@ -26,28 +27,104 @@ This command launches a router connected to your OpenShift master. The router li
+	TemplateRouterConfig templateRouterConfig
+
+	// F5RouterConfig holds the configuration information for f5 router plugin.
+	F5RouterConfig f5RouterConfig
```
I know its a bit too onerous to change how the config is used below + how that structure is passed around. But just a nit - it feels a bit odd to have a config structure with all the different implementing configs inside of it (f5 + template + anything else we add later). Think it would been better to have a specific config for the template router and one specifically for the f5 router and have them reuse (promoted fields) the same base struct containing the common fields (clientcmd config, type and default certificate).

ramr commented on pkg/cmd/infra/router/router.go:
```diff
@@ -26,28 +27,104 @@ This command launches a router connected to your OpenShift master. The router li
+
+	// F5RouterConfig holds the configuration information for f5 router plugin.
+	F5RouterConfig f5RouterConfig
 }
```
Ignore previous comment - too late and not thinking straight.

liggitt replied:
should this and templateRouterConfig be pointers and `nil` when inapplicable?

Miciah commented:
I factored `loadData` out of `loadCert` and `loadKey` and reran the F5 router tests (75.8% coverage) and the router integration tests, which are all still passing.  Does everything look hunky-dory now?

ramr commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,1743 @@
+			// (even though we *ask* for JSON and F5 *usually* obliges).
+			return errorResult
+		}
+		err = decoder.Decode(&errorResult)
```
This doesn't look right -- ```errorResult``` is not JSON.

Miciah replied:
What do you mean? `errorResult` is an `F5Error`, which embeds an `f5Result`, which has the `code` and `message` fields of the JSON response.

ramr replied:
Sorry, what I meant was non 200 http response codes are still valid codes - anything 2xx and 3xx basically can still provide content as per the http spec. So why treat that as an error and not use the response back from the server as the code below does and put it into  ```result``` (if passed) or nil.

Miciah replied:
As far as I recall, I have seen no non-200 responses from F5 BIG-IP that were not errors.  On the other hand, I also have seen no non-4xx responses from F5 BIG-IP other than 200.  I'll change this check to `>= 400`, and it should work just as well and perhaps be clearer.

ramr commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,1743 @@
+			if err != nil {
+				return err
+			}
+		}
```
This check seems redundant as its already done above (line number 626) or does adding an https policy remove the http one?

Miciah replied:
Wow.  Yeah, that's weird.  My next fixup commit will re-arrange the logic to delete the redundant `f5.ensureVserverHasPolicy(f5.httpVserver, httpPolicyName)`.

I'll also move `f5.ensurePolicyExists(httpsPolicyName)`, `f5.ensureDatagroupExists(passthroughRoutesDataGroupName)`, `f5.ensureDatagroupExists(passthroughHostsDataGroupName)`, and `f5.ensureIRuleExists(sslPassthroughIRuleName, sslPassthroughIRule)` outside the `if f5.httpsVserver != ""` block because `addRoute` and other functions assume that the policy and datagroups exist and will act and generate a lot of errors if they do not, so these things needs to exist even if they are not associated with a vserver.

ramr commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,1743 @@
+		Monitor string `json:"monitor"`
+		Name    string `json:"name"`
+	}{
+		Mode:    "round-robin",
```
When @rajatchopra and I looked at the networking bits on F5, there were a few other modes that seemed interesting - fastest, least connections, least sessions and something else (don't recall the name) that used internal counters to make a determination on where to route traffic to. Maybe make this configurable?  Not for now but just something to keep in mind for later.

Miciah replied:
Do you want to extend routes to have a flag for this, or would we only make it configurable on a site-wide basis?

pweil- replied:
One of the things that has been discussed is refactoring the router's internal model to use api objects.  At that point we'd be able to send hints to plugins using annotations that could drive items like this without doing a lot of extensions specific to the underlying implementations.  For instance we could have `openshift.io/f5/balance=round-robin` and `openshift.io/haproxy/balance=leastconn`.

ramr commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,1743 @@
+	members, ok := f5.poolMembers[poolname]
+	if ok {
+		return members, nil
+	}
```
Looks like a local cache - think we should flush this out at some interval otherwise cache staleness could cause issues. Again something that can be done as a follow up.

Miciah replied:
We update F5 at the same time that we update this cache.  There shouldn't be any staleness unless, say, an error occurs while the plug-in is updating F5 BIG-IP (in which case the cache would be more up-to-date) or two instances of the plug-in are running at once (in which case one instance would miss updates from the other instance).

ramr replied:
Couldn't the config be modified via the F5 ui/cmdline?

Miciah replied:
It could.  If that's a use-case we want to support, then we could add a check to invalidate the cache after some period of time has passed, or add a handler for `SIGUSR1` or similar to force a cache-dump (I like this option better), or even do both.

ramr commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,1743 @@
+
+	glog.V(4).Infof("Adding pool member %s to pool %s.", member, poolname)
+
+	url := fmt.Sprintf("https://%s/mgmt/tm/ltm/pool/%s/members",
```
There's a lot of places where these URI paths (well parts of them anyway are generated) are same/similar. In a later PR, maybe see if this can be done in a single place.

Miciah replied:
Just something like a `map[string]string`? That might obfuscate more than improve clarity.  Or maybe something fancier like `map[string]&getResource` where `getResource` would be a function that checked its arguments? That feels slightly over-engineered, but it might e slightly safer.

ramr replied:
What I meant was why not encapsulate all this into a function that just does the URI building part in 1 place - at the very least the URI prefix (as f5LTM has the host name and the ```https://%s/mgmt/tm/ltm/*``` bits could be just in 1 place).

Miciah replied:
OK, something like the following (not tested)?

```
func (f5 *f5LTM) getResourceUrl(path string, args ...interface{}) string {
  return fmt.Sprintf("https://%s/mgmt/tm/" + path, f5.host, args...)
}
```

and then

``` 
  url := f5.getResourceUrl("ltm/pool/%s/members", poolname)
```
 etc.

Note that we have some URLs with `/mgmt/tm/sys/`, so `/mgmt/tm/` is the longest common prefix.

Or would `buildResourceUrl` be better than `get`? Hmm...

ramr commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,1743 @@
+		return err
+	}
+
+	members[member] = true
```
Was this supposed to be a check? ```members``` is a local variable and so doing this is sort of a NOOP.

ramr replied:
Aah probably ```f5.poolMembers[poolname][member] = true```.

Miciah replied:
`f5.poolMembers[poolname]` is a map, so `members` is a reference, not a copy.  We're modifying `f5.poolMembers[poolname]` here.

ramr replied:
Yeah got that after - as always am a bit confused!! ;^)

ramr commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,1743 @@
+	routes, ok := f5.routes[policyname]
+	if ok {
+		return routes, nil
+	}
```
Similar comment to above - local cache - should flush this out at some interval and something for a follow up PR maybe.

ramr commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,1743 @@
+	rulesUrl := fmt.Sprintf("https://%s/mgmt/tm/ltm/policy/%s/rules",
+		f5.host, policyname)
+
+	escapedRulename := f5UrlEscape(routename)
```
minor nit - variable naming ```escapedRouteName``` ... smells like teen spirit err copy-pasta!! ;^)

ramr replied:
Ignore this - my bad - addRoutes threw me off a bit. See that its adding a rule for a route.

Miciah replied:
Yeah, basically a route in OpenShift is implemented using a policy rule in F5 BIG-IP.

ramr commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,1743 @@
+	}{
+		Name: escapedRulename,
+	}
+
```
minor nit here as well for the name ```rulesPayload```.

ramr commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,1743 @@
+		Name: escapedRulename,
+	}
+
+	err := f5.post(rulesUrl, rulesPayload, nil)
```
minor nit here as well for the name ```rulesPayload``` and ```rulesUrl```.

ramr commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,1743 @@
+		Equals:          true,
+		Request:         true,
+		Values:          []string{hostname},
+	}
```
Probably better to have a ```types.go``` file with all the F5 request payload structs defined there. Again something for later.

Miciah replied:
OK, my next fixup commit will put a definition for the unnamed structure type used here for `conditionsPayload` in `types.go` and refer to that type here.

ramr commented on plugins/router/f5/f5.go:
```diff
@@ -0,0 +1,1743 @@
+		Request: true,
+		Select:  true,
+		Vlan:    0,
+	}
```
Similar comment to above - having the struct definition in a separate file improves readability in this function.

Miciah replied:
OK.

ramr commented on plugins/router/f5/plugin.go:
```diff
@@ -0,0 +1,529 @@
+		}
+	}
+
+	for member, _ := range needToDelete {
```
Didn't grok this completely - might need a break as I might just be glossing over it now. 
Don't we need to do the delete pool member before the add if F5 doesn't support the equivalent of a partial PUT?

Miciah replied:
I don't understand you.  We can only manipulate pools by POSTing and DELETEing individual pool members, but `HandleRoute` is given only a list of what's supposed to be in the pool, so what we do here is POST things that should be in the pool but are not and DELETE things that *are* in the pool but should *not* be.

ramr replied:
Aah, I see what I missed - the ```needToDelete[dest] = false```.  Yeah my understanding of the comments above was that we were deleting the existing members and then re-adding them back in. I see now that this is adding the new members in first and then deleting the ones that are missing.

Miciah replied:
I'll try to make the comments clearer.

ramr commented on plugins/router/f5/plugin.go:
```diff
@@ -0,0 +1,529 @@
+
+		// We only delete the pool if the pool is empty, which it may not be
+		// if a service has been added and has not (yet) been deleted.
+		if len(members) <= 0 {
```
I like using the ```<= 0``` but the general convention in this code base is ```== 0```. 
Am partial to putting the constants on the LHS (for an old 'C' hand, lets the compiler catch an assignment vs a comparision bug) but that's moot for golang.

Miciah replied:
Yeah, I like the lvalue-on-RHS for comparisons technique and broader conditional too.  I actually observed the latter in the template router, which is why I went with it here, but if it goes against the wider convention in the OpenShift codebase, I can change this comparison.

ramr commented:
@Miciah finished review - mostly stuff that you can address later. But there was a couple that needed clarifications re: the add pool member and on the delete pool member before the add. 
Other than that and the test failure, it looks LGTM

Miciah commented:
@ramr, thanks! What test failure are you referring to?

ramr commented:
@Miciah  the one that shows up below these comments - 1 failing check -  continuous-integration/travis-ci/pr — The Travis CI build failed.

Miciah commented:
@ramr, OK, I see it now after enabling some JavaScript.  (Also, Travis is working? Neat!) Looks like I need to fix an import for `DeepCopy` somehow...

Miciah commented:
deads pointed me to an API I could use for `DeepCopy`, and I fixed some other issues from Travis as well.  The PR has also been rebased.  The latest code-review fixups commit addresses the following issues that ramr raised:

• `rest_request` now checks the HTTP response to F5 iControl REST requests using `>= 400` rather than `!= 200` to identify error responses.

• `escapedRoutename` and `escapedRulename`, as well as the `f5UrlEscape` function, are gone (they were actually obviated a while ago by a change not to include a route's pathname in the name of the corresponding F5 policy rule).

• `f5.go` is now 100% anonymous-struct free.  All struct types that the F5 router uses for marshalling JSON for F5 iControl REST requests are now in `types.go`, and are now better commented too (bonus!).

• Functions in `f5.go` are more thoroughly commented.

• `len(foo) <= 0` has been changed to `len(foo) == 0`.

• Comments in `updatePool` in `plugin.go` should be clearer now.

• The logic in `Initialize` has been fixed.

Travis made me fix the following:

• Some formatting issues have been fixed (Travis must have a newer version of `gofmt` than I have).

• The F5 router tests were fixed to call `httptest.Server.Close()` when finished with the httptest server.  Without this, Travis was timing out in `TestF5RouterSuccessiveInstances` (timed out after 1 minute—although the tests were taking under 5 seconds total on my system).

• The Bash completions have been updated.

The last item also required a rebase in order to resolve merge conflicts.

Miciah commented:
Travis is still randomly failing tests because it times out making requests to the httptest webserver that is set up in some of the F5 router testcases.  I've spent some time trying to get the tests working in Travis, but deads suggests just tagging the PR with [test], so let's try that...

Miciah commented:
Got a weird error:

```
+ vagrant origin-init --stage inst --os fedora origin_test_pull_requests_origin_3799
Usage: vagrant [options] <command> [<args>]
```

[test] again.

Miciah commented:
The F5-router tests succeeded using Jenkins (in just under 1 minute):

```
ok  	github.com/openshift/origin/plugins/router/f5	52.519s	coverage: 75.6% of statements
```

The integration tests failed because of the websocket issue I mentioned above:

```
Running TestRouter...

failed  TestRouter
Started, serving at 10.97.145.38:8080
Started, serving at 10.97.145.38:8888
Started, serving TLS at 10.97.145.38:8443
Stopping listener at 10.97.145.38:8080
Stopping listener at 10.97.145.38:8888
Stopping listener at 10.97.145.38:8443
Server message: accept tcp 10.97.145.38:8443: use of closed network connectionServer message: accept tcp 10.97.145.38:8080: use of closed network connectionStopping listener at 10.97.145.38:8080
Stopping listener at 10.97.145.38:8888
Stopping listener at 10.97.145.38:8443
Server message: accept tcp 10.97.145.38:8888: use of closed network connectionStopping listener at 10.97.145.38:8080
Stopping listener at 10.97.145.38:8888
--- FAIL: TestRouter-2 (61.60s)
	router_test.go:261: Unable to verify response: websocket.Dial ws://www.example.com:80/echo: bad status
	router_test.go:265: TC websocket unsecure failed! Response body  did not match expected hello-websocket-unsecure
	router_test.go:261: Unable to verify response: websocket.Dial wss://www.example.com:443/echo: bad status
	router_test.go:265: TC ws edge termination failed! Response body  did not match expected hello-websocket-edge
	router_test.go:261: Unable to verify response: websocket.Dial wss://www.example.com:443/echo: bad status
	router_test.go:265: TC ws passthrough termination failed! Response body  did not match expected hello-websocket-passthrough
FAIL
Running TestRouterDuplications...

ok      TestRouterDuplications
Running TestRouterPathSpecificity...

ok      TestRouterPathSpecificity
```

@pweil-, @ramr, how shall I proceed?

ramr commented:
@Miciah not sure - are the f5 tests using the same ports (```8080/8888/8443```) - that causing a conflict? 
You can run the integration tests locally and check ( ```hack/test-integration.sh```).

pweil- commented:
Yes, I would run these tests locally (`test-integration-docker.sh`).  You can then comment out the teardown or pause the test if you need to and check out the containers.  It looks like it is failing on websocket dial.  

Possibly https://github.com/pweil-/origin/blob/6d11b07a1d87fbdfeb579ecab6a763c1d464dbb4/test/integration/router_test.go#L538-538 needs changed.

Miciah commented:
@ramr, @pweil-, it's an issue with Golang's websocket client.  I mentioned the problem in https://github.com/openshift/origin/pull/2940#issuecomment-122426821.  The integration tests run fine in a devenv with an entry in `/etc/hosts` for www.example.com.

pweil- commented:
@Miciah @ramr - you are digging up a long lost memory now!  The websocket client is indeed whack :-)  That's why the websocket code was setting the header in a funky way I believe.  In fact, I brought this up in the golang-nuts forum but didn't get much feedback: https://groups.google.com/forum/#!searchin/golang-nuts/websocket$20header/golang-nuts/AOdjWHTi0ps/PpIfc5ny4IkJ

My preference would be to leave the working websocket code in so that the haproxy tests work during a normal run since that is our reference implementation.

Miciah commented:
@pweil-, OK, I'll revert the changes to the websocket tests and instead add some comments around the problem code.

Miciah commented:
I took a different approach rather than reverting the changes: https://github.com/Miciah/origin/commit/4f7aaec83d55fcbe604e2d35aa50e3145f8f5aa3 adds a workaround that changes the route alias to 0.0.0.0 if the default router image (viz. the template router) is used, and prints out a message about `/etc/hosts` if another image is used and a testcase fails.  This way, the hackery is special-cased.  That is, the testcases are defined using reasonable values, and then those values are changed as needed for the template-router case.

@ramr, is this approach OK?

Miciah commented:
To be clear, I did run `hack/test-integration-docker.sh` both with the template router (with and without www.example.com in `/etc/hosts`) and with the F5 router (with www.example.com in `/etc/hosts`), and it succeeds now in all cases.

liggitt commented on pkg/cmd/admin/router/router.go:
```diff
@@ -121,6 +185,13 @@ func NewCmdRouter(f *clientcmd.Factory, parentName, name string, out io.Writer)
+	cmd.Flags().StringVar(&cfg.ExternalHostPassword, "external-host-password", cfg.ExternalHostPassword, "If the underlying router implementation connects with an external host, this is the password for authenticating with the external host.")
+	cmd.Flags().StringVar(&cfg.ExternalHostHttpVserver, "external-host-http-vserver", cfg.ExternalHostHttpVserver, "If the underlying router implementation uses virtual servers, this is the name of the virtual server for HTTP connections.")
+	cmd.Flags().StringVar(&cfg.ExternalHostHttpsVserver, "external-host-https-vserver", cfg.ExternalHostHttpsVserver, "If the underlying router implementation uses virtual servers, this is the name of the virtual server for HTTPS connections.")
+	cmd.Flags().StringVar(&cfg.ExternalHostPrivateKey, "external-host-privkey", cfg.ExternalHostPrivateKey, "If the underlying router implementation requires an SSH private key, this is the path to the private key file.")
```
`external-host-private-key`

Miciah replied:
I have fixed this in an fixups commit.

liggitt commented on pkg/cmd/admin/router/router.go:
```diff
@@ -259,28 +342,40 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg *
+			"ROUTER_EXTERNAL_HOST_HTTP_VSERVER":  cfg.ExternalHostHttpVserver,
+			"ROUTER_EXTERNAL_HOST_HTTPS_VSERVER": cfg.ExternalHostHttpsVserver,
+			"ROUTER_EXTERNAL_HOST_INSECURE":      strconv.FormatBool(cfg.ExternalHostInsecure),
+			"ROUTER_EXTERNAL_HOST_PRIVKEY":       privateKey,
```
-1, we are working to get rid of secrets in envvars, not expand them. The existing cert/key in an envvar will go away as soon as possible.

Miciah replied:
So `$ROUTER_EXTERNAL_HOST_PRIVKEY` should instead contain a filename, and that means we need a volume config to bind-mount `$ROUTER_EXTERNAL_HOST_PRIVKEY` into the router container to some fixed location, right?

Miciah replied:
I have fixed this in an upcoming fixups commit.  I also had to update the integration tests to do the same bind-mount thing.

liggitt replied:
Envvar can be a file path, and you should create a secret with the contents of the private key and a secret volume mount

Miciah replied:
I'm not familiar with secrets and secret volume mounts.  Is there a good example I could reference?

liggitt replied:
cli/secrets/new.go illustrates creating a secret:
https://github.com/openshift/origin/blob/master/pkg/cmd/cli/secrets/new.go#L241-L245
https://github.com/openshift/origin/blob/master/pkg/cmd/cli/secrets/new.go#L167

cli/cmd/volume.go creates a secret volume source and volume mount:
https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/volume.go#L394-L396
https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/volume.go#L438-L442

liggitt commented on pkg/cmd/infra/router/router.go:
```diff
@@ -27,28 +28,104 @@ This command launches a router connected to your cluster master. The router list
+	Config *clientcmd.Config
+
+	// Type is the type of plugin being created: f5 or template.
+	Type string
```
if you make the configs pointers, you don't need this type, you can just key off whether a particular config is nil or not

Miciah replied:
I'm having difficulty understanding how this would work.  My thinking goes as follows:

1. In `NewCommandRouter`, declare a temporary variable `foo` before the Cobra `Run` callback, so that it will close over the variable.

2. Use `flag.StringVar(type, "type", ...` instead of using `cfg.Type`.

3. Defer the other flag bindings until we know what the type flag's value is.

Is Step 3 possible with Cobra? We are maybe binding flags to things that may or may not exist, with those things' existences' being contingent on the value of the type flag.  How do we conditionalise some flag bindings on the value of another flag?

liggitt replied:
I mean each plugin has a function that adds bindings to a flagset. You'd call each function, passing in the flagset, to build the command

Miciah replied:
I understand that part, but the bindings would be to some `cfg.TemplateRouterConfig.foo` or some `cfg.F5RouterConfig.foo`, and we don't know whether to create `cfg.TemplateRouterConfig` or `cfg.F5RouterConfig` until we have read the type flag, which happens after flags are bound, right?

Miciah replied:
Unless you want to create `cfg.TemplateRouterConfig` and `cfg.F5RouterConfig` both and delete one or the other after the fact.  Is that what you meant?

liggitt replied:
Sure, create both up front, bind args to both, then use one or the other in the router config. The separate config objects, bind functions, and validation are sort of similar to the way we have master and node arg objects that both know how to bind args, which we reuse in the all-in-one case.

liggitt commented on pkg/cmd/infra/router/router.go:
```diff
@@ -27,28 +28,104 @@ This command launches a router connected to your cluster master. The router list
-	ReloadScript       string
+// routerConfig represents a container for all other router config types.
+// Any global config belongs on this level.
+type routerConfig struct {
```
f5 and template bits should probably be split into their own files:
1. config struct
2. flag bindings
3. `make*Plugin` function

Miciah replied:
I'll hold off on the refactoring effort here until I understand how you mean for the flag bindings to be handled.  Is this refactoring perhaps deferrable?

Miciah commented:
The latest fixups commit makes the following changes:

• The `oadm` command's flag `-external-host-privkey` has been changed to `-external-host-private-key`.

• `oadm` now stores the path to the SSH private key, rather than storing the key itself, in the `ROUTER_EXTERNAL_HOST_PRIVKEY` environment variable for the `openshift-router` command.

• `openshift-router` has a new flag: `-f5-private-key`, which can be used as an alternative to setting `ROUTER_EXTERNAL_HOST_PRIVKEY`.

• `oadm` now generates a Secret object containing the SSH private key, which `oadm` adds to the service account, as well as a Volume object containing the Secret and a VolumeMount for the Volume, with which `oadm` configures the router's container  (see `generateSecretsConfig`).

• The default file mode for files in secrets volume is 0444, it is impossible to override that default file mode (see https://github.com/kubernetes/kubernetes/issues/4789), and secrets volumes are read-only, so in order for the `oadm` use-case to work properly, it was necessary to change the F5 router to copy the private key file into a new file and chmod that file with permissions acceptable to the `ssh` and `scp` commands.

• A while back, the `oadm` had been modified to configure a liveness probe for the router, which is reasonable for the HAProxy router, but in the case of the F5 router (which does not respond to HTTP requests), it was causing the router to be restarted repeatedly, so the liveness probe has been conditionalised on the template router type (see `generateLivenessProbeConfig`).

• `f5RouterConfig` and `makeF5Plugin` have been moved from `pkg/cmd/infra/router/router.go` to the new file `pkg/cmd/infra/router/f5.go` alongside the new `bindFlagsForF5RouterConfig` func, which has been factored out of `NewCommandRouter`.

• `templateRouterConfig` and `makeTemplatePlugin` have been moved from `pkg/cmd/infra/router/router.go` to the new file `pkg/cmd/infra/router/template.go` alongside the new `bindFlagsForTemplateRouterConfig` func, which has been factored out of `NewCommandRouter`.

• `f5RouterConfig` now contains `*f5RouterConfig` and `*templateRouterConfig` instead of `f5RouterConfig` and `templateRouterConfig`.  `NewCommandRouter` now initialises these pointers before invoking Cobra, and its `Run` callback now zeroes whichever pointer contains the configuration object that is irrelevant to the specified router type.

• The F5 router now runs the `ssh` and `scp` commands with the `-o GSSAPIAuthentication=no` and `-o PubkeyAuthentication=yes` flags for added robustness.

• `HandleRoute` now calls `ensurePoolExists` even if the route is only being modified in case the router is told to update a route that didn't already exist.

• A few commends and log messages have been fixed.

I also rebased and adjusted for some API changes.

Miciah commented:
Note for testing: `oadm` now creates a Secret object for the F5 router, but deleting the router's DeploymentConfig, ReplicationController, and Pod does not delete the Secret; it must be cleaned up manually (`oc delete secrets/external-host-private-key-secret`).

Miciah commented:
@liggitt, @ramr, does this look OK?

ramr commented:
@Miciah  will review this tmrw/over the weekend.

ramr commented:
@Miciah  - there's some potential follow up items here but we can address those once this PR gets in. 
Saw you addressed the govet build issues, so this LGTM  :shipit:

Miciah commented:
Terrific! Would you like me to squash everything into one commit for merging?

rajatchopra commented:
@Miciah Yes please squash the commits to a logical few please. Thanks. The jenkins failure looks unreal meanwhile.

Miciah commented:
How would you like it split up? If you don't want it all in a single commit, I could do one commit with changes strictly related to the template router (the refactoring and comments clean-up around the template router code and data structures that came up during the reviews here) and then one commit to add the F5 router, or I could try to make it more fine-grained if you have suggestions.

Miciah commented:
[test] again because the failure looks like a provisioning issue.

rajatchopra commented:
@miciah one commit is fine too if it is a big hassle to have two or three. A split between templates and the router seems just fine.

Miciah commented:
Squish, squash, applesauce.

rajatchopra commented:
[merge]

Miciah commented:
Looks like the test failure is from a timeout connecting to the registry in `pkg/generate/app/cmd`, not related to the changes in this PR.  [test] again.

rajatchopra commented:
re [merge]

Miciah commented:
Yay!

pweil- commented:
:beers: :tada:

smarterclayton commented:
Something I hit while working in here.

The router command for F5 should not be the same as the infra router -
there is no reason a template router and an F5 router to have the same
basic arguments or types (they do need common filtering, but the remainder
of their config is disjoint).  I'll handle that as part of the work I'm
doing for scoping routes, but in the future we should be sure to only add
new options to commands when their behavior is different.  Miciah, I'll cc
you on the pull requests that splits them out.


On Fri, Aug 28, 2015 at 11:27 AM, Paul Weil <notifications@github.com>
wrote:

> [image: :beers:] [image: :tada:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/2940#issuecomment-135788004>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

smarterclayton commented:
The F5 router command should be using a secret to store its passwords, not
just the command line and env.

On Sun, Aug 30, 2015 at 4:50 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> Something I hit while working in here.
>
> The router command for F5 should not be the same as the infra router -
> there is no reason a template router and an F5 router to have the same
> basic arguments or types (they do need common filtering, but the remainder
> of their config is disjoint).  I'll handle that as part of the work I'm
> doing for scoping routes, but in the future we should be sure to only add
> new options to commands when their behavior is different.  Miciah, I'll cc
> you on the pull requests that splits them out.
>
>
> On Fri, Aug 28, 2015 at 11:27 AM, Paul Weil <notifications@github.com>
> wrote:
>
>> [image: :beers:] [image: :tada:]
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/2940#issuecomment-135788004>.
>>
>
>
>
> --
> Clayton Coleman | Lead Engineer, OpenShift
>



-- 
Clayton Coleman | Lead Engineer, OpenShift
